### PR TITLE
[Feature] Add support for custom metadata providers through a REST API

### PR DESCRIPTION
**Pull request by Dewyer**
# About
This MR is a WiP followup for this issue: https://github.com/advplyr/audiobookshelf/issues/2467

I am trying to implement the provider in ABS. This is my first time contributing to the project so, sorry I am sure I didn't do things exactly as I should, LMK what should I change to better align with the coding style of the project.

## Notes
- I am aware that I am missing a bucnh of transaltion items, thats WiP
- Custom providers should probably get their own page on the guides site, I will at some point write an MR for that

### CODE REVIEW

**Comment by Dewyer** [comment]
I am trying to fully test it myself, but probably will miss a lot of edge cases.

File: server/Database.js [hunk 1]
```diff
@@ -694,6 +700,45 @@ class Database {
...
+   */
+  async removeCustomMetadataProviderById(id) {
+    // destroy metadta provider
+    await this.customMetadataProviderModel.destroy({
+      where: {
+        id,
+      }
+    })
+
+    const slug = `custom-${id}`;
```
Comment (advplyr) [inline server/Database.js:730]: Is there a reason to use a slug here instead of just the UUID?

File: client/strings/en-us.json [hunk 2]
```diff
@@ -96,6 +96,7 @@
   "HeaderAudiobookTools": "Audiobook File Management Tools",
   "HeaderAudioTracks": "Audio Tracks",
   "HeaderAuthentication": "Authentication",
+  "HeaderCustomMetadataProviders": "Custom metadata providers",
```
Comment (advplyr) [inline client/strings/en-us.json:99]: These should be alphabetical (ignoring casing). Also, don't worry about adding the strings to the other languages, I have a local script I run for that.

File: server/controllers/MiscController.js [hunk 3]
```diff
@@ -717,5 +717,95 @@ class MiscController {
...
+   * @param {import('express').Request} req
+   * @param {import('express').Response} res
+   */
+  async addCustomMetadataProviders(req, res) {
+    if (!req.user.isAdminOrUp) {
+      Logger.error(`[MiscController] Non-admin user "${req.user.username}" attempted to add admin custom metadata providers`)
+      return res.sendStatus(403)
+    }
+
+    const { name, url, apiKey } = req.body;
```
Comment (advplyr) [inline server/controllers/MiscController.js:766]:
```suggestion
    const { name, url, apiKey } = req.body
```
The repo is not using semicolons and not using trailing commas. I have this rule setup for vscode but still working on improvements for this #2183

File: server/controllers/MiscController.js [hunk 4]
```diff
@@ -717,5 +717,95 @@ class MiscController {
...
+    if (!id) {
+      return res.status(500).send(`Invalid delete data`)
+    }
+
+    const provider = await Database.customMetadataProviderModel.findByPk(id);
+    await Database.removeCustomMetadataProviderById(id);
+
+    SocketAuthority.adminEmitter('custom_metadata_provider_removed', provider)
+
+    res.json({})
```
Comment (advplyr) [inline server/controllers/MiscController.js:808]:
```suggestion
    res.sendStatus(200)
```

File: server/models/CustomMetadataProvider.js [hunk 5]
```diff
@@ -0,0 +1,58 @@
...
+    }
+  }
+
+  static findByPk(id) {
+    return this.findOne({
+      where: {
+        id,
+      }
+    })
+  }
```
Comment (advplyr) [inline server/models/CustomMetadataProvider.js:35]:
```suggestion
```

This function is built-in to Sequelize. https://sequelize.org/docs/v6/core-concepts/model-querying-finders/#findbypk

File: server/models/CustomMetadataProvider.js [hunk 6]
```diff
@@ -0,0 +1,58 @@
...
+   * @param {import('../Database').sequelize} sequelize 
+   */
+  static init(sequelize) {
+    super.init({
+      id: {
+        type: DataTypes.UUID,
+        defaultValue: DataTypes.UUIDV4,
+        primaryKey: true
+      },
+      name: DataTypes.STRING,
```
Comment (advplyr) [inline server/models/CustomMetadataProvider.js:48]:
```suggestion
      mediaType: DataTypes.STRING,
      name: DataTypes.STRING,
```
We'll want to support podcast custom metadata providers as well in the future or any other media types that could get added.

File: server/providers/CustomProviderAdapter.js [hunk 7]
```diff
@@ -0,0 +1,70 @@
...
+        const providerId = providerSlug.split("custom-")[1]
+        const provider = await Database.customMetadataProviderModel.findByPk(providerId);
+
+        if (!provider) {
+            throw new Error("Custom provider not found for the given id");
+        }
+
+        const matches = await axios.get(`${provider.url}/search?query=${encodeURIComponent(title)}${!!author ? `&author=${encodeURIComponent(author)}` : ""}`, {
+            headers: {
+                "Authorization": provider.apiKey,
```
Comment (advplyr) [inline server/providers/CustomProviderAdapter.js:19]: If a user is using bearer token authorization or basic authorization would they be instructed to enter that in the `apiKey` input? Let me know if that doesn't make sense

File: custom-metadata-provider-specification.yaml [hunk 8]
```diff
@@ -0,0 +1,124 @@
...
+          type: string
+        subtitle:
+          type: string
+        author:
+          type: string
+        narrator:
+          type: string
+        publisher:
+          type: string
+        published_year:
```
Comment (advplyr) [inline custom-metadata-provider-specification.yaml:89]: Should we use camelCase here to stay consistent with our data models? publishedYear

File: client/pages/config/custom-metadata-providers.vue [hunk 9]
```diff
@@ -0,0 +1,45 @@
...
+    <app-settings-content :header-text="$strings.HeaderCustomMetadataProviders">
+      <template #header-items>
+        <ui-tooltip :text="$strings.LabelClickForMoreInfo" class="inline-flex ml-2">
+          <a href="https://www.audiobookshelf.org/guides/#" target="_blank" class="inline-flex">
+            <span class="material-icons text-xl w-5 text-gray-200">help_outline</span>
+          </a>
+        </ui-tooltip>
+        <div class="flex-grow" />
+
+        <ui-btn color="primary" small @click="setShowAddModal()">{{ $strings.ButtonAdd }}</ui-btn>
```
Comment (advplyr) [inline client/pages/config/custom-metadata-providers.vue:12]:
```suggestion
        <ui-btn color="primary" small @click="setShowAddModal">{{ $strings.ButtonAdd }}</ui-btn>
```

File: server/providers/CustomProviderAdapter.js (hunk 7 above)
Comment (Dewyer) [reply server/providers/CustomProviderAdapter.js:19]: Basic auth isn't supported in the current model, as it would require 2 pieces of credentials, but bearer, is essentially just Authorization header + a Bearer prefix and an api key so they could make that work

File: server/Database.js (hunk 1 above)
Comment (Dewyer) [reply server/Database.js:730]: it makes it easier to know a provider is a custom one, as they all have this prefix, if it was just an id you could technically have some future provider be prefixed with valid hex characters or something

**Comment by FlyinPancake** [comment]
I have addressed most of the concerns, however I held out on the `MediaType` since I don't know what values should be stored there. 

As of now and according to my knowledge, all the providers are assumed to be book / audiobook metadata providers.

I think one CMP should support multiple `MediaType`s and those could be stored in a comma separated `string`, making it extensible in the future. 

Supporting podcasts would require a separate CMP EP in `custom-metadata-provider-specification.yaml`. We could come up with something, but I would like to keep this PR to the point and addressing it later.

We could rename the `/search` EP to `/searchBooks` to accommodate future expandability.

File: server/providers/CustomProviderAdapter.js (hunk 7 above)
Comment (advplyr) [reply server/providers/CustomProviderAdapter.js:19]: Basic auth would be supported in this model as far as I understand it because the user is required to enter the full authorization header value.
`Basic <credentials>` or `Bearer <api key>`.

My question was about if we are requiring the user to write in the `Bearer` part of the authorization header. If we are then `apiKey` might not be an accurate name for the variable.

**Comment by advplyr** [comment]
I made a lot of updates to get this ready. You will need to delete the table from your db before using the updates. 

In the model I added new properties and changed `apiKey` to `authHeaderValue`. It's technically not just an api key going there but the full authorization header value. I also made this not required which seems reasonable if someone makes a public API. The openapi spec may need to be updated to account for that.

I moved the config page to go under "Item Metadata Utils" instead of being in the main sidebar.

I updated the library api request to return the custom metadata providers so there is no longer a need for a client-only api endpoint to get custom providers. The socket requests have been updated to send all clients the client-safe JSON.

I created a new controller and set the stage to eventually allow custom podcast providers. The `url` is validated to make sure it is a real url.

I want to add more validation to the data being returned from the custom metadata provider. Please test your provider on the edge container when you get a chance.

**Comment by FlyinPancake** [comment]
Thanks for doing this ❤️

Ur a legend


I'll get around to it as soon as I can

**Comment by FlyinPancake** [comment]
Hi! 

I checked the edge container out, it works as expected.
This PR does not call searchBooks it calls search it that is a problem, this is the time to act. 

As this is a merged pr, i'll create the example custom metadata provider and probably a guide for the website if that's ok

**Comment by advplyr** [comment]
I realized a better solution was to pass the `mediaType` to the `/search` endpoint as a query param.

I just made that update here: https://github.com/advplyr/audiobookshelf/commit/4e3e7b10ce70042d2ffa9e391cb18af6de9f8377

So if in the future someone implements a provider for both books and podcasts they can check the `mediaType` passed in.

If you can start on the guide that would be great. I would like to have something up before this goes in the next release.
